### ISSUE TITLE
[FEATURE] add metrics for drift comparisons

### ISSUE THREAD

**Issue by koaning** [koaning/scikit-lego#266]
There is a need for metrics that can detect drift. 

```
drift_metric(X, X_) 
```

There are multiple ways of doing this, but we should start with a few. Some inspiration could be found over on [the scipy stack](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.wasserstein_distance.html).

**Comment by koaning** [comment]
Here's a heuristic approach I might want to propose. 

![image](https://user-images.githubusercontent.com/1019791/73611840-68b97280-45e6-11ea-86dd-8bc90ed99281.png)

**Comment by koaning** [comment]
My only caveat here is that this idea only works if `X_orig` and `X_new` are of equal length. We might be able to ignore that by doing (bootstrap-style) sampling.

**Comment by koaning** [comment]
@MBrouns does this feel like a worthwhile starter?

**Comment by MBrouns** [comment]
Good to get started on something like this! 

Question: Do you have an idea on how using wasserstein on the likelihoods compares to only taking the likelihood of L_new given d_orig?

**Comment by MBrouns** [comment]
Also, something that might be relevant to know for this: there are interesting probabilistic data structures for on-line quantile accumulation (https://github.com/tdunning/t-digest)

**Comment by koaning** [comment]
I've checked the [python variant](https://github.com/CamDavidsonPilon/tdigest) but it only seems to update single dimensional data structures. This will work on a single column, but not an entire dataset.

**Comment by koaning** [comment]
> Do you have an idea on how using wasserstein on the likelihoods compares to only taking the likelihood of L_new given d_orig?

I have one likelihood per row in each dataset. I need a method of comparing those. Wasserstein seems to have properties that we like.

**Comment by MBrouns** [comment]
Wouldn't comparing the (bootstrapped) likelihood of the entire set work? e.g. take n samples from both sets, sum the logp's for all observations in that sample and see if the likelihood of X_new is higher than X_orig.

**Comment by koaning** [comment]
There's a property about earth-movers-distance that I wouldn't want to ignore. 

Let's suppose we compare these two distributions; 

![image](https://user-images.githubusercontent.com/1019791/73614686-a8da1e80-4601-11ea-80f0-bbe428efebd1.png)

The original dataset is marked. A direct comparison of likelihoods would suggest overlap and wassersteins distance would be modest. 

Now imagine that they drift apart. 

![image](https://user-images.githubusercontent.com/1019791/73614704-ce672800-4601-11ea-8bc9-cd45536a066b.png)

The likelihood comparison will receive the highest value *possible* and wasserstein will be higher than before. 

Now imagine that the distributions differ even more. 

![image](https://user-images.githubusercontent.com/1019791/73614722-edfe5080-4601-11ea-8f78-b8f915674aec.png)

The number from our likelihood approach remains one. The wasserstein distance will increase still. This is a property I like. 

## Notes on the Approach 

![image](https://user-images.githubusercontent.com/1019791/73614733-066e6b00-4602-11ea-8bd2-aad060def153.png)

My proposal has two parts that require a method; 

- how we estimate the distribution of the original data (potentially this receives a param)
- how the take the likelihoods of `L_old` and `L_new` and compare them (potentially this receives a param)

**Comment by koaning** [comment]
And yes. This is me figuring out what remote communication will be like for this project.

**Comment by koaning** [comment]
It might help if we agree on the signature of a drift metric. The more I think about it, the more I feel it should be an object. 

```python
drift = DriftDetector(param_a, param_b)
drift.fit(X)
drift.score_drift(X_new) # number comes out
```

The nice thing is that this might also allow us to use pipelines. 

```python
from sklearn.preprocessing import QuantileTranformer 
from sklearn.pipeline import Pipeline 

pipe = Pipeline([
    ("qt", QuantileTranformer()),
    ("dd", DriftDetector(param_a, param_b))
])
```

Another benefit is that we can start thinking about drift detection techniques that also use the label. 

```python
pipe.fit(X)
pipe.fit(X, y)
```

In the end though I think that all the techniques that we can come up with are but mere flags. They can indicate something is happening but to properly understand what is changing you'll need to go indepth.

**Comment by koaning** [comment]
If this idea has any merit. It should be its own package.
